pkg/client/selfhosted/selfhosted.go

```diff
@@ -0,0 +1,232 @@
+	lookupURL   = "%s/%s/%s/tags/list?n=500"
+	manifestURL = "%s/%s/%s/manifests/"
+)
+
+type Options struct {
+	URL      string
+	LoginURL string
+	Username string
+	Password string
+	Bearer   string
+}
+
+type Client struct {
+	*http.Client
+	Options
+}
+
+type AuthResponse struct {
+	Token string `json:"token"`
+}
+
+type TagResponse struct {
+	Name string   `json:"name"`
+	Tags []string `json:"tags"`
+}
+
+type Result struct {
+	Name      string  `json:"name"`
+	Timestamp string  `json:"last_updated"`
+	Images    []Image `json:"images"`
+}
+
+type Image struct {
+	Digest       string `json:"digest"`
+	OS           string `json:"os"`
+	Architecture string `json:"Architecture"`
+}
+
+type ManifestResponse struct {
+	SchemaVersion int        `json:"schemaVersion"`
+	MediaType     string     `json:"mediaType"`
+	Manifests     []Manifest `json:"manifests"`
+}
+
+type Manifest struct {
+	MediaType string   `json:"mediaType"`
+	Size      int      `json:"size"`
+	Digest    string   `json:"digest"`
+	Platform  Platform `json:"platform"`
+}
+
+type Platform struct {
+	Architecture string `json:"architecture"`
+	Os           string `json:"os"`
+}
+
+func New(ctx context.Context, opts Options) (*Client, error) {
+	client := &http.Client{
+		Timeout: time.Second * 5,
+	}
+
+	// Only try to setup auth if an actually URL is present.
+	if opts.URL != "" {
+		// Setup Auth if username and password used.
+		if len(opts.Username) > 0 || len(opts.Password) > 0 {
+			if len(opts.Bearer) > 0 {
+				return nil, errors.New("cannot specify Bearer as well as username/password")
+			}
+
+			token, err := basicAuthSetup(client, opts)
+			if err != nil {
+				return nil, fmt.Errorf("failed to setup auth: %s", err)
+			}
+			opts.Bearer = token
+		}
+	}
+
+	return &Client{
+		Options: opts,
+		Client:  client,
+	}, nil
+}
+
+func (c *Client) Tags(ctx context.Context, _, repo, image string) ([]api.ImageTag, error) {
+	url := fmt.Sprintf(lookupURL, c.Options.URL, repo, image)
+	urlManifest := fmt.Sprintf(manifestURL, c.Options.URL, repo, image)
+	var tags []api.ImageTag
+
+	response, err := c.doRequest(ctx, url)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, tag := range response.Tags {
+
+		manifestResponse, err := c.doManifestRequest(ctx, urlManifest+tag)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, manifest := range manifestResponse.Manifests {
+
+			sha := strings.Replace(manifest.Digest, "sha256:", "", 1)
+
+			tags = append(tags, api.ImageTag{
+				Tag:          tag,
+				SHA:          sha,
+				Timestamp:    time.Now(),
+				OS:           manifest.Platform.Os,
+				Architecture: manifest.Platform.Architecture,
+			})
+		}
+	}
+
+	return tags, nil
+}
+
+func (c *Client) doManifestRequest(ctx context.Context, url string) (*ManifestResponse, error) {
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.URL.Scheme = "https"
```

Comment:
Can we remove enforcing the https scheme?

I run local registries which serve on http, and this means it will never work. Can we not enforce, and instead enforce a scheme on the config? That way people will make a conscious decision to write `http` rather than `https`, whilst not dropping the http use case.

Reply:
Agreed. This is now removed. Perhaps a new PR should consist of adding some form of enforcing, depending on how you want to deal with URL's in general.